[Go] Add generator and acceptance tests

garimakhulbe opened the pull request:
* Added beta version go generator
* Added acceptance tests
 - boolean 
 - integer 
 - number (minus decimal)
 - string (minus base64url)
 - file and formdata tests
 - array (minus UnixTime)
 - dictionary (minus UnixTime)
 - date, datetime, datetimerfc1123
 - duration
 - complex (minus polymorphic and polyrecursice)
 - url
 - http
 - model-flattening
 - validation
 - required-optional
 - customBaseUri

azuresdkci commented:
Can one of the admins verify this patch?

azurecla commented:
Hi __@garimakhulbe__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (gaagar). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

garimakhulbe commented:
This PR adds go generator and go acceptance tests in AutoRest.

In current list of 521 acceptance tests, 18 tests are failing. Below are things I'm not sure about....
- Below tests are failing because in current go generator, I have added an empty check for required parameters and it throws validation error at client when required param is empty. I followed most of validation logic from C# generator but I could not find empty check or might implement differently. I have added this check in generator because we have [issue](https://github.com/Azure/azure-sdk-for-go/issues/316) in Go SDK where user was sending empty in path parameter and getting 4xx. After this change, user will get validation error with name of empty parameter. I think we can ignore these tests in go coverage. Not sure.
```
 UrlQueriesStringEmpty, UrlPathsByteEmpty, putArrayEmpty, putDictionaryEmpty, putStringEmpty,
 UrlPathsStringEmpty, UrlQueriesByteEmpty, UrlQueriesArrayCsvEmpty
```
- These optional coverage tests are failing because go does not support redirect for 307 Put, Delete, Post and Patch and 300 Get and Head. We can ignore these tests in go coverage.
```
HttpRedirect307Put, HttpRedirect307Post, HttpRedirect307Patch, HttpRedirect307Delete,
HttpRedirect300Get, HttpRedirect300Head
```
- ```putComplexReadOnlyPropertyValid``` - for swagger read only property, I have added validation check at client which checks for readonly null and throws validation error.
- ```putComplexArrayValid``` - server expects one of the property as null string however go can't have string as null so failing. Either we need to remove null check from server or remove this test from coverage. 
- ```UrlQueriesArrayCsvEmpty``` - as per go generator design, empty query never gets added to request if its not a required param, however server expects empty query for this test and it fails. I think we can remove this test as optional empty query will never send in request. 

- TODO: Tests left in coverage for implementation - Complex polymorphic and polymorphic recursive tests, Base64Url, UnixTime, Decimal tests.

Few more points to add...
* TODO: Go generator have not implemented all the extensions yet and it breaks for few swaggers for now: ContainerService, Graphrbac and Datalake/catalog. We are working on these and will add them in coming PRs.
* Go has one generator currently and does not have AutoRest.Go.Azure flavor currently. 
* Go generator uses [go-autorest](https://github.com/Azure/go-autorest) as a helper library in generated code. We use [glide.lock](https://github.com/Azure/azure-sdk-for-go/blob/master/glide.lock#L5) file in Go SDK for managing dependencies. Every new SDK version has its glide.lock with dependencies commit ids SDK tied too.

This PR on behalf of me and Mariana(@mcardosos).

Reviewers and other interested folks: @brendandixon @vishrutshah @eduardkoller

vishrutshah commented:
@tbombach, @fearthecowboy & @John-Hart : Could you please have a quick look at the general structure, gulp changes etc.. of the Go language merging into AutoRest? 

@mcardoso: Could you please have a quick look at the Go generator? 

Thanks,

tbombach commented:
@azuresdkci add to whitelist

tbombach commented:
Really awesome work! Gulp and travis changes look good (I'll look at the general generator structure and tests as well).

We have to get Go on the Jenkins machines as well so that those builds pass as well. Our (AutoRest core) team needs to add commands to [initialize-settings.ps](https://github.com/Azure/autorest/blob/master/Tools/initialize-settings.ps1) (so that anyone can set up their machine for Go), then run it on all of the build machines. I'm working on modifying the PS script right now, then I can send a pull request to your repository with the changes.

garimakhulbe commented:
@tbombach Thanks for helping me. Build is failing for glide package. We need [glide](https://github.com/Masterminds/glide) package to run tests and need to add below command in ps script. 

After installing go and setting gopath.
_go get -u github.com/Masterminds/glide_

cc: @mcardoso

vishrutshah commented on src/generator/AutoRest.Go/GoCodeGenerator.cs:
```diff
@@ -0,0 +1,116 @@
+        {
+            // TODO resource string.
+            get { return "Go Client Libraries"; }
+        }
```
Azure specific Go code generator.

vishrutshah commented on src/generator/AutoRest.Go/GoCodeGenerator.cs:
```diff
@@ -0,0 +1,116 @@
+                Settings.Namespace + ".Models");
+        }
+
+        private void PopulateAdditionalProperties(ServiceClient serviceClientModel)
```
>PopulateAdditionalProperties [](start = 21, length = 28)

we should remove this if not used or provide more comments why is it commented out.

garimakhulbe replied:
Removed.

vishrutshah commented:
We should update cli documentation with go information @ [https://github.com/Azure/autorest/blob/94b1a21667d8a393af90cefdc47e0585811cad0d/Documentation/cli.md](https://github.com/Azure/autorest/blob/94b1a21667d8a393af90cefdc47e0585811cad0d/Documentation/cli.md)

vishrutshah commented on src/core/AutoRest/AutoRest.json:
```diff
@@ -35,6 +35,9 @@
     "AzureResourceSchema": {
       "type": "AzureResourceSchemaCodeGenerator, AutoRest.AzureResourceSchema"
+    },
+    "Go": {
```
>Go [](start = 5, length = 2)

Should this be `Azure.Go` or does it support generating generic Go SDK generation?

garimakhulbe replied:
Whats the difference in generic and Azure.go? We only use it for generating Azure Go SDK ARM code.

garimakhulbe replied:
@vishrutshah Sorry I was wrong earlier. I discussed with Brendan, this is a generic generator.

vishrutshah replied:
Does this generator support x-ms-odata, x-ms-pageable, x-ms-long-running-operations etc... If so i believe that this is not generic and should be named as "Azure.Go"

---
In reply to: [74860338](https://github.com/Azure/autorest/pull/1336#discussion_r74860338) [](ancestors = 74860338)

vishrutshah commented on src/generator/AutoRest.Go/Extensions.cs:
```diff
@@ -0,0 +1,898 @@
+        }
+
+        public static string Capitalize(this string value)
+        {
```
minor: should have summary of the methods

vishrutshah commented on src/generator/AutoRest.Go/rakefile.rb:
```diff
@@ -0,0 +1,231 @@
+require 'fileutils'
+
+AUTOREST     = "../../core/AutoRest/bin/Debug/net451/win7-x64/AutoRest.exe"
```
>AUTOREST [](start = 0, length = 8)

could you please elaborate the purpose of this (Ruby) file in Go?

mcardosos replied:
That file, as far as I understand, when run it regenerates the ARM stuff for the SDK

garimakhulbe replied:
We use this to generate Go SDK. I will comments in rakefile.

garimakhulbe replied:
This is done.

vishrutshah commented on src/generator/AutoRest.Go/TemplateModels/ServiceClientTemplateModel.cs:
```diff
@@ -0,0 +1,134 @@
+                            ? string.Empty
+                            : packageName;
+
+            // TODO (gosdk): Derive client name from Swagger once the Swagger settles down.
```
>// TODO (gosdk): Derive client name from Swagger once the Swagger settles down [](start = 12, length = 78)

can we uncomment this now or what are the blockers?

garimakhulbe replied:
Removed this comment as this is not relevant now. This was for the ManagementClient name which we have kept same in all packages.

```Go
// ManagementClient is the base client for Batch.
type ManagementClient struct {
	autorest.Client
	BaseURI        string
	APIVersion     string
	SubscriptionID string
}
```

vishrutshah commented on src/generator/AutoRest.Go/Templates/VersionTemplate.cshtml:
```diff
@@ -0,0 +1,47 @@
+const (
+    major = "3"
+    minor = "1"
+    patch = "0"
```
I see the latest release being 3.2.0-beta?

garimakhulbe replied:
Since its in review, I used the old stable branch to create SDK.

vishrutshah commented on src/generator/AutoRest.Go.Tests/src/tests/glide.lock:
```diff
@@ -0,0 +1,19 @@
+hash: e33d98082bdab9f30032ec6f8f9e2539e549ccbe3dfaf45f1a16702217018912
+updated: 2016-08-10T17:03:36.0631005-07:00
+imports:
+- name: github.com/Azure/go-autorest
```
Do we need to check-in the lock file or it should be ignored and regenerated everytime?

garimakhulbe replied:
Yes, I will remove this lock file.

garimakhulbe replied:
I removed the lock file.

tbombach commented:
@garimakhulbe I created a PR to your branch: https://github.com/garimakhulbe/autorest/pull/2 which added go to our setup PS script for Windows. I used it to set up my machine and was able to run `gulp test:go`

We still need to run it on our build machines though. @cormacpayne I'm OOF tomorrow - could you run my new version of the `initialize-settings.ps1` script on the builder machines? It's located in this branch: https://github.com/tbombach/autorest/blob/go-setup/Tools/initialize-settings.ps1

garimakhulbe commented:
Thanks @tbombach for updating the script. I will coordinate with @cormacpayne on this. Thanks again.

garimakhulbe commented:
Adding more folks from the team @ahmetalpbalkan @paulmey @boumenot @colemickens @anandramakrishna

tbombach commented:
@azuresdkci retest this please!

tbombach commented:
@garimakhulbe Jenkins build passes now!: http://azuresdkci.cloudapp.net/job/autorest/3073/

Travis seems to be failing during a couple of the Go tests: https://travis-ci.org/Azure/autorest/builds/151913484, but I'm not sure why. It doesn't seem related to the initialize-settings changes, since it's failing on a network request to the test server

```
Error: report.ManagementClient#GetReport: Failure sending request: StatusCode=0 -- Original Error: Get http://localhost:3000/report: dial tcp 127.0.0.1:3000: getsockopt: connection refused
    stderr: panic: runtime error: invalid memory address or nil pointer dereference
```

and

```
body-formdata_test.go:32:
    c.Assert(err, chk.IsNil)
... value autorest.DetailedError = autorest.DetailedError{Original:(*url.Error)(0xc8201221b0), PackageType:"formdatagroup.FormdataClient", Method:"UploadFileViaBody", StatusCode:0, Message:"Failure sending request"} ("formdatagroup.FormdataClient#UploadFileViaBody: Failure sending request: StatusCode=0 -- Original Error: Put http://localhost:3000/formdata/stream/uploadfile: EOF")
```

I'm not exactly sure how we can fix this. Your latest pushed commit restarted tests, so we'll see if those builds pass.

garimakhulbe commented:
@tbombach Yes, it is not related to script. It is related to my code and will update these failing tests. Thanks a lot for your help!

garimakhulbe commented:
@vishrutshah comments were addressed. Let me know if it looks good now.

vishrutshah commented on gulpfile.js:
```diff
@@ -485,6 +513,19 @@ gulp.task('regenerate:expected:csazurecomposite', function (cb) {
+    'mappings': goMappings,
+    'outputDir': 'src/tests/generated',
+    'codeGenerator': 'Go',
+    'flatteningThreshold': '1'
```
>flatteningThreshold [](start = 5, length = 19)

Does Go generator really supports flattening threshold? or was it just copied from other gulp task?

garimakhulbe replied:
@mcardoso do you know why we use this here?

mcardosos replied:
I just copied :(

garimakhulbe replied:
@vishrut What is ```flatteningThreshold```? We can remove it if not implemented. I know its one of argument of Autorest command but we never use it code generator.

vishrutshah commented:
looks like by mistake you checked-in src/generator/AutoRest.Go.Tests/src/tests/acceptancetests/sample.png or is it necessary?

vishrutshah commented on src/generator/AutoRest.Go.Tests/src/tests/generated/body-byte/bytegroupoperations.go:
```diff
@@ -0,0 +1,287 @@
+package bytegroup
```
>package [](start = 0, length = 7)

Aren't the generated files gitignored? Why do i see them in the last commits or they are different?

garimakhulbe replied:
I am confused whether to put the generated code in PR or not. I have included the generated tests in case if someone wants to refer to them but not sure what is the right way. Let me know if these should be removed.

mcardosos commented:
sample.png is usted for form data tests. Node also has it checked in (and it looks like python tests use the Node sample.png too). It would be cuter if we all used the same image, but not one checked in with tests.

garimakhulbe commented:
We are using this file in Formdata test. Even though these tests are not included in coverage this time, I am still keeping them and this file here for coming PRs.

mcardosos commented:
Tiny observation, some files still have the Apache license

garimakhulbe commented:
@mcardosos which files?

mcardosos commented:
Extensions, IScopeProvider, MethodScopeProvider, Template, VariableScopeProvider, and everything under TemplateModels

garimakhulbe commented:
@mcardosos I checked all other generator are using MIT License. Will update to..

> // Licensed under the MIT License. See License.txt in the project root for license information.

garimakhulbe commented:
License text is fixed.

garimakhulbe commented:
@vishrutshah @mcardosos All comments are addressed. Please let me know if it looks good now. We want this merge soon.

vishrutshah commented:
@garimakhulbe Changes looks good. Please let us know once you have thumbs up from your team (who is expert on Go) as they know the in & out of generator. After that we can merge this in! Thanks!

garimakhulbe commented:
Thanks @vishrutshah for comments. My team is fine with the generator code. Please merge. 
cc @brendandixon

vishrutshah commented:
@garimakhulbe @brendandixon Merged Go generator into AutoRest master. Thanks!

tbombach commented:
Congratulations! Getting the Go generator merged and tests passing is a huge accomplishment. Really great work!

garimakhulbe commented:
Thanks @vishrutshah @mcardosos @tbombach @brendandixon for your help and comments.
